server/src/com/thoughtworks/go/server/service/GoConfigService.java

```diff
@@ -581,17 +585,33 @@ public void updateAgentAttributes(String uuid, String userName, String hostname,
             command.addCommand(GoConfigDao.updateApprovalStatus(uuid, false));
         }
 
-        if (enable.isFalse()){
+        if (enable.isFalse()) {
             command.addCommand(GoConfigDao.updateApprovalStatus(uuid, true));
         }
 
         if (hostname != null) {
             command.addCommand(new GoConfigDao.UpdateAgentHostname(uuid, hostname, userName));
         }
+
         if (resources != null) {
             command.addCommand(new GoConfigDao.UpdateResourcesCommand(uuid, new Resources(resources)));
         }
 
+        if (environments != null) {
+            Set<String> existingEnvironments = cruiseConfig().getEnvironments().environmentsForAgent(uuid);
+            Set<String> newEnvironments = new HashSet<>(Arrays.asList(environments.split(",")));
+
+            Set<String> environmentsToRemove = Sets.difference(existingEnvironments, newEnvironments);
+            Set<String> environmentsToAdd = Sets.difference(newEnvironments, existingEnvironments);
+
+            for (String environmentToRemove : environmentsToRemove) {
+                command.addCommand(new GoConfigDao.ModifyEnvironmentCommand(uuid, environmentToRemove, TriStateSelection.Action.remove));
+            }
+
+            for (String environmentToAdd : environmentsToAdd) {
+                command.addCommand(new GoConfigDao.ModifyEnvironmentCommand(uuid, environmentToAdd, TriStateSelection.Action.add));
```

Comment:
This can fail and throw an exception, if the environment doesn't exist. I'm not sure that will be handled gracefully. I don't see any test for that.